core/drivers/zynqmp_csu_puf.c

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Foundries.io Ltd.
+ *
+ * Jorge Ramirez-Ortiz <jorge@foundries.io>
+ */
+
+#include <initcall.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <mm/core_memprot.h>
+
+#include "drivers/zynqmp_csu.h"
+#include "drivers/zynqmp_csu_puf.h"
+
+#define PUF_CMD_OFFSET			0x00
+#define PUF_CFG0_OFFSET			0x04
+#define PUF_CFG1_OFFSET			0x08
+#define PUF_SHUT_OFFSET			0x0C
+#define PUF_STATUS_OFFSET		0x10
+#define PUF_WORD_OFFSET			0x18
+
+#define PUF_REGISTRATION		1
+#define PUF_REGENERATION		4
+#define PUF_STOP			3
+#define PUF_RESET			6
+
+#define PUF_CFG0_DEFAULT		0x02
+#define PUF_CFG1_DEFAULT		0x0c230090
+#define PUF_SHUT_DEFAULT		0x01000100
+#define PUF_REGEN_TIME_MS		3000
+#define PUF_STATUS_SYN_WRD_RDY_TIMEOUT	500000
+#define PUF_STATUS_SYN_WRD_RDY_MASK	BIT(0)
+#define PUF_STATUS_KEY_RDY_MASK		BIT(3)
+
+#define	PUF_SYN_WORDS			141
+
+TEE_Result zynqmp_csu_puf_regenerate(void)
+{
+	vaddr_t puf = core_mmu_get_va(ZYNQMP_CSU_PUF_BASE, MEM_AREA_IO_SEC,
+				      ZYNQMP_CSU_PUF_SIZE);
+	vaddr_t csu = core_mmu_get_va(CSU_BASE, MEM_AREA_IO_SEC, CSU_SIZE);
+	uint32_t status = 0;
+
+	if (!puf || !csu)
+		return TEE_ERROR_GENERIC;
+
+	io_write32(puf + PUF_CFG0_OFFSET, PUF_CFG0_DEFAULT);
+	io_write32(puf + PUF_SHUT_OFFSET, PUF_SHUT_DEFAULT);
+	io_write32(puf + PUF_CMD_OFFSET, PUF_REGENERATION);
+	mdelay(PUF_REGEN_TIME_MS);
+
+	status = io_read32(csu + ZYNQMP_CSU_ISR_OFFSET);
+	if (status & ZYNQMP_CSU_ISR_PUF_ACC_ERROR_MASK) {
+		EMSG("regeneration failed");
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+void zynqmp_csu_puf_reset(void)
+{
+	vaddr_t puf = core_mmu_get_va(ZYNQMP_CSU_PUF_BASE, MEM_AREA_IO_SEC,
+				      ZYNQMP_CSU_PUF_SIZE);
+
+	io_write32(puf + PUF_CMD_OFFSET, PUF_RESET);
+}
+
+/* TODO:
+ *  the syndrome words need to be post-processed and fused to generate a
+ *  persistent KEK
+ */
+static uint32_t syn[PUF_SYN_WORDS];
+
+static TEE_Result zynqmp_csu_puf_init(void)
+{
+	vaddr_t puf = core_mmu_get_va(ZYNQMP_CSU_PUF_BASE, MEM_AREA_IO_SEC,
+				      ZYNQMP_CSU_PUF_SIZE);
+	uint64_t tref = timeout_init_us(PUF_STATUS_SYN_WRD_RDY_TIMEOUT);
+	uint32_t status = 0;
+	uint32_t index = 0;
```

Comment:
`index` is a reserved function name, please use another name.

Reply:
ok